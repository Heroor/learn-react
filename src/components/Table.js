import React from 'react'
import PropTypes from 'prop-types'
import Button from './Button'

const largeColumn = {
  width: '50%'
}
const midColumn = {
  width: '20%'
}
const smallColumn = {
  width: '10%'
}

const Table = ({list, remove}) =>
  <ul className="table">
    {list.map(v => (
      <li key={v.objectID} className="table-row">
        <span style={smallColumn}>
          {v.points}
        </span>
        <span style={largeColumn}>
          <a
            target="blank"
            href={v.url}
          >
            {v.title}
          </a>
        </span>
        <span style={midColumn}>
          {v.author}
        </span>
        <span style={smallColumn}>
          {v.num_comments}
        </span>
        <span style={smallColumn}>
          <Button
            className="button-inline"
            onClick={() => remove(v)}
          >
            [DEL]
          </Button>
        </span>
      </li>
    ))}
  </ul>


// filter callback
// const filterList = key => v =>
//   v.title.toLowerCase().includes(key.toLowerCase())

Table.propTypes = {
  list: PropTypes.arrayOf(
    // 每一项元素的内容：
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  ).isRequired,
  remove: PropTypes.func.isRequired
}

export default Table
