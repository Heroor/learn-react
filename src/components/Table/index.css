.table {
  margin: 20px 0;
}

.table-header {
  display: flex;
  line-height: 24px;
  font-size: 16px;
  padding: 0 10px;
  justify-content: space-between;
}

.table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
}

.table-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  background: #ffffff;
  border: 1px solid #e3e3e3;
  border-radius: 4px;
  box-shadow: 0 0 8px rgba(7, 17, 27, .1);
}

.table-header>span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.table-row>span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}