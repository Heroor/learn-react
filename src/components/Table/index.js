import React, {Component} from 'react'
import {sortBy} from 'lodash'
import classnames from 'classnames'

import PropTypes from 'prop-types'
import Button from '../Button'

const largeColumn = {
  width: '50%'
}
const midColumn = {
  width: '20%'
}
const smallColumn = {
  width: '10%'
}

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'point')
}


const Sort = ({
  sortKey,
  activeSortKey,
  onSort,
  children
}) => {
  const sortClass = classnames(
    'button-inline',
    {'button-active': sortKey === activeSortKey}
  )

  return (
    <Button
      onClick={() => onSort(sortKey)}
      className={sortClass}
    >
      {children}
    </Button>
  )
}

class Table extends Component {
  constructor(props) {
    super(props)
    this.state = {
      sortKey: 'NONE',
      isSortReverse: false
    }
    this.onSort = this.onSort.bind(this)
  }

  onSort(sortKey) {
    this.setState(prevState => {
      const isSortReverse = prevState.sortKey === sortKey && !prevState.isSortReverse
      return {
        sortKey,
        isSortReverse
      }
    })
  }

  render() {
    const {
      list,
      remove
    } = this.props
    const {
      sortKey,
      isSortReverse
    } = this.state
    const onSort = this.onSort
    const sortedList = SORTS[sortKey](list)
    const reverseSortedList = isSortReverse ?
      sortedList.reverse() :
      sortedList
    return (
      <div className="table">
        <div className="table-header">
          <span style={smallColumn}>
            <Sort
              sortKey={'POINTS'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Points
            </Sort>
          </span>
          <span style={largeColumn}>
            <Sort
              sortKey={'TITLE'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Title
            </Sort>
          </span>
          <span style={midColumn}>
            <Sort
              sortKey={'AUTHOR'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Author
            </Sort>
          </span>
          <span style={smallColumn}>
            <Sort
              sortKey={'COMMENTS'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Comments
            </Sort>
          </span>
          <span style={smallColumn}>
            <Sort>
              Archive
            </Sort>
          </span>
        </div>

        <ul>
          {reverseSortedList.map(v => (
            <li key={v.objectID} className="table-row">
              <span style={smallColumn}>
                {v.points}
              </span>
              <span style={largeColumn}>
                <a
                  target="blank"
                  href={v.url}
                >
                  {v.title}
                </a>
              </span>
              <span style={midColumn}>
                {v.author}
              </span>
              <span style={smallColumn}>
                {v.num_comments}
              </span>
              <span style={smallColumn}>
                <Button
                  className="button-inline"
                  onClick={() => remove(v)}
                >
                  [DEL]
                </Button>
              </span>
            </li>
          ))}
        </ul>
      </div>
    )
  }
}


// filter callback
// const filterList = key => v =>
//   v.title.toLowerCase().includes(key.toLowerCase())

Table.propTypes = {
  list: PropTypes.arrayOf(
    // 每一项元素的内容：
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  ).isRequired,
  activeSortKey: PropTypes.string,
  remove: PropTypes.func.isRequired
}

Sort.proptypes = {
  sortKey: PropTypes.string,
  onSort: PropTypes.func,
  children: PropTypes.node
}

Sort.defaultProps = {
  onSort: () => {}
}

export default Table
