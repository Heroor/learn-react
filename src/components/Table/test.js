import React from 'react'
import ReactDOM from 'react-dom'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderComponentToSnapshot, { hitList } from '../../config/TestConfig'
import Table from './index'

Enzyme.configure({
  adapter: new Adapter()
})

// Table test
describe('Table', () => {
  const props = {
    list: hitList,
    remove() {
      console.log('remove!')
    }
  }

  it('render without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Table {...props} />, div)
  })

  test('has a calid snapshot', () => {
    renderComponentToSnapshot(<Table {...props} />)
  })

  // 单元测试
  //Enzyme API 中总共有三种渲染机制。你已经知道了 shallow() ，这里还有 mount() 和
  // render() 方法。这两种方式都会初始化父组件和所有的子组件。此外 mount() 还给予
  // 你调用组件生命周期的方法。但是什时候该使用哪种渲染机制呢？这里有一些建议：
  // • 不论怎样都优先尝试使用浅渲染（ shallow() ）
  // • 如果需要测试 componentDidMount() 或 componentDidUpdate() ，使用 mount()
  // • 如果你想测试组件的生命周期和子组件的行为，使用 mount()
  // • 如果你想测试一个组件的子组件的渲染，并且不关心生命周期方法和减少些渲染的
  // 花销的话，使用 render()
  it('shows two items in list', () => {
    const element = shallow(
      <Table {...props} />
    )
    expect(element.find('.table-row').length).toBe(2)
  })
})