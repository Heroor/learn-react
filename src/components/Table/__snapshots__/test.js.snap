// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has a calid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Archive
      </button>
    </span>
  </div>
  <ul>
    <li
      className="table-row"
    >
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      />
      <span
        style={
          Object {
            "width": "50%",
          }
        }
      >
        <a
          href={undefined}
          target="blank"
        >
          1
        </a>
      </span>
      <span
        style={
          Object {
            "width": "20%",
          }
        }
      >
        1
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        1
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          [DEL]
        </button>
      </span>
    </li>
    <li
      className="table-row"
    >
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      />
      <span
        style={
          Object {
            "width": "50%",
          }
        }
      >
        <a
          href={undefined}
          target="blank"
        >
          2
        </a>
      </span>
      <span
        style={
          Object {
            "width": "20%",
          }
        }
      >
        2
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        2
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          [DEL]
        </button>
      </span>
    </li>
  </ul>
</div>
`;
